openedx/core/djangoapps/password_policy/compliance.py

```diff
@@ -0,0 +1,140 @@
+"""
+Utilities for enforcing and tracking compliance with password policy rules.
+"""
+from datetime import datetime
+
+import pytz
+from dateutil.parser import parse as parse_date
+from django.conf import settings
+from django.utils.translation import ugettext as _
+
+from util.date_utils import DEFAULT_SHORT_DATE_FORMAT, strftime_localized
+from util.password_policy_validators import validate_password
+
+
+class NonCompliantPasswordException(Exception):
+    """
+    Exception that should be raised when a user who is required to be compliant with password policy requirements
+    is found to have a non-compliant password.
+    """
+    pass
+
+
+class NonCompliantPasswordWarning(Exception):
+    """
+    Exception that should be raised when a user who will soon be required to be compliant with password policy
+    requirements is found to have a non-compliant password.
+    """
+    pass
+
+
+def should_enforce_compliance_on_login():
+    """
+    Returns a boolean indicating whether or not password policy compliance should be enforced on login.
+    """
+    config = _rollout_config()
+    return config.get('ENFORCE_COMPLIANCE_ON_LOGIN', False)
+
+
+def enforce_compliance_on_login(user, password):
+    """
+    Verify that the user's password is compliant with password policy rules and determine what should be done
+    if it is not.
+
+    Raises NonCompliantPasswordException when the password is found to be non-compliant and the compliance deadline
+    for the user has been reached. In this case, login should be prevented.
+
+    Raises NonCompliantPasswordWarning when the password is found to be non-compliant and the compliance deadline for
+    the user is in the future.
+
+    Returns None when the password is found to be compliant, or when no deadline for compliance has been set for the
+    user.
+
+    Important: This method should only be called AFTER the user has been authenticated.
+    """
+    is_compliant = _check_user_compliance(user, password)
+    if is_compliant:
+        return
+
+    deadline = _get_compliance_deadline_for_user(user)
+    if deadline is None:
+        return
```

Comment:
Would it make more sense to do this check first, BEFORE running validations? In theory, we should only need to run validations for users that have a deadline for compliance. This check involves a DB query to check the courseaccessroles table though, so it may actually be more efficient to leave things the way they are.

Reply:
I agree that logically, the checks should be reversed. But also agreed that performance-wise, the current ordering is good. I'd leave as-is.

Reply:
We talked about this in person, but to follow up -- now that this branch doesn't check PasswordHistory, there are no db hits when calling _check_compliance, so no problems doing that first and _get_compliance_deadline second.